src/irr/asset/CImageWriterJPG.cpp

```diff
@@ -98,16 +96,25 @@ static void jpeg_file_dest(j_compress_ptr cinfo, io::IWriteFile* file)
 	dest->file = file;
 }
 
-
-#ifndef NEW_SHADERS
 /* write_JPEG_memory: store JPEG compressed image into memory.
 */
-static bool writeJPEGFile(io::IWriteFile* file, const asset::CImageData* image, uint32_t quality)
+static bool writeJPEGFile(io::IWriteFile* file, const const asset::ICPUImage* image, uint32_t quality)
 {
-	auto format = image->getColorFormat();
+	const auto& imageParams = image->getCreationParameters();
+	const auto& region = image->getRegions().begin();
+	auto format = imageParams.format;
+
+	IImage::SBufferCopy::TexelBlockInfo blockInfo(format);
+	core::vector3du32_SIMD trueExtent = IImage::SBufferCopy::TexelsToBlocks(region->getTexelStrides(), blockInfo);
+
 	bool grayscale = (format == asset::EF_R8_SRGB) || (format == asset::EF_R8_UNORM);
 	
-	core::vector3d<uint32_t> dim = image->getSize();
+	core::vector3d<uint32_t> dim;
+	dim.X = trueExtent.X;
+	dim.Y = trueExtent.Y;
+	dim.Z = trueExtent.Z;
+
+	const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(format) * trueExtent.X;
```

Comment:
this is gonna end badly for multi-region images.

Best course of action is to create a new image with only one region, one layer and one mip-map level

then if input image is EF_R8_SRGB or EF_R8G8B8_SRGB run a copy filter into the new image.
if not then use a:
- `CConvertFormatImageFilter<EF_UNKNOWN,EF_R8G8B8_SRGB>` for input images with more than 1 channel
- `CConvertFormatImageFilter<EF_UNKNOWN,EF_R8_SRGB>` for input images with just 1 channel
(whenever inFormat or outFormat of `CConvertFormatImageFilter` is EF_UNKNOWN then the filter uses the non-templated runtime conversion variant by looking up the inFormat or outFormat from the inImage or outImage creation parameters)

This way, no matter what happens you'll have a clean Single Region, Single Mip-map image to save.

Reply:
the tip with EF_UNKOWN has been added to docs